Spigot-API-Patches/0243-Added-SoundBuilder.patch

```diff
@@ -0,0 +1,331 @@
++    @NotNull
++    SoundBuilder receivers(@Nullable Player... receivers);
++
++    /**
++     * Selects all players within a cuboid selection around the sound location, within the
++     * specified bounding box. If you want a more spherical check, see {@link #receivers(int,
++     * boolean)}
++     *
++     * @param radius amount to add on all axis
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder receivers(int radius);
++
++    /**
++     * Selects all players within the specified radius around the sound location. If byDistance is
++     * false, behavior uses cuboid selection the same as {@link #receivers(int, int)} If byDistance is
++     * true, radius is tested by distance in a spherical shape
++     *
++     * @param radius amount to add on each axis
++     * @param byDistance true to use a spherical radius, false to use a cuboid
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder receivers(int radius, boolean byDistance);
++
++    /**
++     * Selects all players within a cuboid selection around the sound location, within the
++     * specified bounding box. Allows specifying a different Y size than X and Z If you want a more
++     * cylinder check, see {@link #receivers(int, int, boolean)} If you want a more spherical check,
++     * see {@link #receivers(int, boolean)}
++     *
++     * @param xzRadius amount to add on the x/z axis
++     * @param yRadius amount to add on the y axis
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder receivers(int xzRadius, int yRadius);
++
++    /**
++     * Selects all players within the specified radius around the sound. If byDistance is
++     * false, behavior uses cuboid selection the same as {@link #receivers(int, int)} If byDistance is
++     * true, radius is tested by distance on the y plane and on the x/z plane, in a cylinder shape.
++     *
++     * @param xzRadius amount to add on the x/z axis
++     * @param yRadius amount to add on the y axis
++     * @param byDistance true to use a cylinder shape, false to use cuboid
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder receivers(int xzRadius, int yRadius, boolean byDistance);
++
++    /**
++     * Selects all players within a cuboid selection around the sound location, within the
++     * specified bounding box. If you want a more cylinder check, see {@link #receivers(int, int,
++     * boolean)} If you want a more spherical check, see {@link #receivers(int, boolean)}
++     *
++     * @param xRadius amount to add on the x axis
++     * @param yRadius amount to add on the y axis
++     * @param zRadius amount to add on the z axis
++     * @return a reference to this object.
++     */
++    @NotNull
++    SoundBuilder receivers(int xRadius, int yRadius, int zRadius);
++
++    /**
++     * @return Location of where the sound will play
++     */
++    @Nullable
++    Location location();
```

Comment:
What does a null return value signify? Why is it allowed?

Reply:
because at this point, location CAN be null

Reply:
Seems like a design error... Builder instance that has a "build" method but isn't guaranteed to have all of its required fields set.

Reply:
build method? No, everything that is allowed to be non-zero but could be zero during the build is checked in the play method. Also, you can't set a null as location but only NotNull, so this is just a backup for a non-set value.

Reply:
In another comment thread, you accepted to make locations mandatory, therefore they are non-null now.